### ISSUE TITLE
Default the null-safe runtime strong-ness of an entrypoint by it's language version

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/sdk#41203]
This is a meta issue covering several work items

See the "blocked by" list

**Comment by sigmundch** [comment]
I'm a bit confused - I thought the consensus was that external tools (bazel/build_runner) would determine whether the app is mixed/weak or not, and tell the tools what to do with the flag, but that it was not the role of the tools to manage the default anymore. Did this change?

**Comment by kevmoo** [comment]
> I'm a bit confused - I thought the consensus was that external tools (bazel/build_runner) would determine whether the app is mixed/weak or not, and tell the tools what to do with the flag, but that it was not the role of the tools to manage the default anymore. Did this change?

@johnniwinther @natebosch @jakemac53 @leafpetersen ?

**Comment by natebosch** [comment]
In the cases where the user is interacting with these tools through a build system, this will indeed be handled by the build system.

These issues specifically cover the paces where users interact with the command line directly. Specifically if someone calls `dart bin/main.dart` or `dart2js web/main.dart` we want the mode to be determined by the language version in that `main.dart` library.

**Comment by lrhn** [comment]
Is there a design document for this? These issues appear completely void of context.

So, if a user writes a *new* app package, not putting any SDK constraint into the pubspec.yaml and therefore getting the newest language version for the package, if any package they depend on is not Null Safe, then the current package's main will still trigger strong mode?
Or will it only do that if all the libraries in the program are actually null safe?

**Comment by sigmundch** [comment]
I'm only aware of the discussion that was happening under this language PR https://github.com/dart-lang/language/pull/779

**Comment by jakemac53** [comment]
> So, if a user writes a _new_ app package, not putting any SDK constraint into the pubspec.yaml and therefore getting the newest language version for the package, if any package they depend on is not Null Safe, then the current package's main will still trigger strong mode?

Yes, and then since we are no longer allowing running with strong null safety on in mixed mode this should give a compile time error. This is somewhat orthogonal though.

> Or will it only do that if all the libraries in the program are actually null safe?

This would be prohibitively expensive - it would require reading all files in the program before even beginning the compile.

Imo it would also be more surprising/confusing to the user. It is much easier to reason about the single entrypoint file for your app than all its transitive deps. Adding/removing dependencies would start changing the mode you are running in by default etc.

**Comment by lrhn** [comment]
I'm sure it can be made to work, but it does mean that a new user will trip over this on their first program.

When we release 2.10.0, a new user downloads the new SDK and starts writing a new program.
They depend on some package which hasn't migrated yet (there'll be plenty of those), and then their program will fail to compile. We need to be *very* clear in the messaging on that compilation error.
The solution is to add the magic incantation `//@dart=2.9` to their "main" file. (Or rather, make a new main file with that, and have it call the real Null Safe main file). That's almost completely undiscoverable, so we need to say that in bold letters, before and after any other error message (do we have flashing text?), and with plenty of IDE support.

The design of language versioning was made so that if you do nothing, you get the most recent language version for your code, and all other packages gets what they were configured for.
This choice makes it a *problem* to get the most recent language version for your main file. That's not an optimal situation.

I'm not saying I have a good solution to the problem, but this solution use data intended for one purpose (select the language version for a single library) for a secondary purpose (force entire program to be NNBD), and that's always dangerous and fragile.

**Comment by jakemac53** [comment]
Yes - I definitely agree that it isn't a perfect solution but it is the best solution I have seen proposed for determining what the default should be.

You are also overstating the problem a bit I think. It is only a problem _if_ the user has some (transitive) deps that aren't opted in (+ null safe).

**Comment by alexmarkov** [comment]
Running application in null-safe mode should be conscious decision made by user because it changes the semantics of the program considerably. Even when all libraries are opted-in and statically clean it is likely that users will see certain number of errors at run time when they first try to run their application in null-safe mode. Those errors could appear in their dependencies, not necessary in their own code. As a result, many users will have to override such default.

Automatically enabling null-safe mode if main library is opted-in also means that users will not be able to opt-in and migrate their main package until the rest of the application and all dependencies are migrated (unless they override the mode).

**Comment by jakemac53** [comment]
Longer term it should be very rare for people to want to run in weak mode. That is meant as a transitionary thing only, and it provides very little value other than allowing you to run. You lose almost all safety provided by nnbd by running in this mode.

For that reason it does not make sense to run in weak mode by default. We should still provide a mechanism to opt _out_  of strong mode, but the default should be strong if the entrypoint is opted in. This is a clear signal from the user they they want strong null safety.

**Comment by natebosch** [comment]
> Running application in null-safe mode should be conscious decision made by user because it changes the semantics of the program considerably.

Forcing an opt-in flag will be less safe in the long run. Anyone who doesn't know about the flag won't use it, they'll think they have correct code and publish but then others won't be able to use it in strong mode. Anyone who knows to use it locally but forgets to add it to their `.travis.yml` file will risk introducing bugs for strong mode down the line.

Defaulting based on the entrypoint will not break existing packages - you have to opt in an existing package by changing the SDK constraint.

Defaulting based on the entrypoint will be confusing for current users who will need to learn the additional semantics of the language version of the library with `main()`. This is a real problem, I think it's less drastic than the problems with other approaches. In the long term it's also what we _want_ new user experience to be. When I download the SDK and try out Dart I _want_ as much safety as possible. The problem only exists during some transition period where most of the ecosystem isn't migrated yet. This is also mitigated by the fact that we can use this same criteria consistently across all tools the user interacts with. They don't need to learn different flag or configuration styles to run in strong mode across different tools. Once you learn these semantics once, you can apply them anywhere and it won't be surprising.

> Automatically enabling null-safe mode if main library is opted-in also means that users will not be able to opt-in and migrate their main package until the rest of the application and all dependencies are migrated (unless they override the mode).

The majority of entrypoint libraries should be very small. It should not be a problem that this library can't be migrated until everything else is, because the migration there is small and straightforward.

In the rare case where the entrypoint library is large and the user wants to migrate it before everything works all it takes is 1 extra library to be the new entrypoint and call the `main()` from an import.

**Comment by alexmarkov** [comment]
I agree that using strong mode by default makes sense longer term, after the most of ecosystem is migrated and packages are actually tested in null-safe mode. Eventually we will remove weak mode entirely.

However, for now people have not even started migrating their applications and only a few packages are migrated, so weak mode is still useful. While running in weak mode users can migrate and opt-in their code more or less independently. After cleaning all static error and opting-in, they can take another step of enabling null-safety at run time. That step may need to do more changes in their application, and even more important, in their dependencies. That's why I think it should not happen automatically.

> The majority of entrypoint libraries should be very small. It should not be a problem that this library can't be migrated until everything else is, because the migration there is small and straightforward.

Correct me if I'm wrong, but I thought the recommended way of opting in is to specify SDK constraint _per package_. So, unless we advertise magical `//@dart=2.xx` comments, users will have to opt-out the whole package which contains their main library.

**Comment by natebosch** [comment]
> I agree that using strong mode by default makes sense longer term, after the most of ecosystem is migrated and packages are actually tested in null-safe mode.

This can't happen until Dart 3.0
I _really_ don't want to have an unsafe default until we're ready for a breaking SDK release for which we have no schedule or plan.

> unless we advertise magical `//@dart=2.xx` comments, users will have to opt-out the whole package which contains their main library.

I think we will have to advertise these.

**Comment by jakemac53** [comment]
Note also that I don't think we are really encouraging out of order migrations in general - they are somewhat unsafe and you could easily get broken by a dependency opting into nnbd but not doing a breaking change release of their package. It is not entirely obvious that regardless of how you do the migration opting in to nnbd is a breaking change (even many people on the Dart team don't/didn't realize this).

**Comment by leafpetersen** [comment]
@lrhn I'm quite surprised that you're resistant to this, it's directly in line with your comments [here](https://github.com/dart-lang/language/issues/771#issuecomment-573567040).

In any case, I strongly feel that this is the correct solution, and was what I proposed (based on discussions with @mit-mit) in [this PR](https://github.com/dart-lang/language/pull/779) that I think most of the people on this thread were looped in on.  

> Or will it only do that if all the libraries in the program are actually null safe?

This is the worst possible UX.  It means that when some random library in my deps opts in (that happens to be the last non-null safe library in my deps), all of the sudden there's a silent transition to strong mode **even though I changed nothing**.

> The design of language versioning was made so that if you do nothing, you get the most recent language version for your code, and all other packages gets what they were configured for.

That is **precisely** what this aims to arrange for.  If you do nothing, you get the most recent language version for your code, and one of two things happens:
  - Some package you depend on has not migrated, and you get told "You have unmigrated upstream dependencies, you should opt your package out.   If you really want to be opted in, you have to pass this special --no-null-safety flag and you're on your own".
  - All packages you depend on are migrated, and you get run in the language version you asked for.  The packages you depend on, since they are migrated, should also be expecting to be run in strong mode (since their tests will be auto-opted as well).

It's true that we can't know for certain that an upstream migrated package works for strong mode, but we also can't know that it works with weak mode.  So we have to pick one or the other (we're not going to support running different pieces of the app in different modes).  This approach makes sure that for folks who stay on the blessed path (migrate in order) everything just works.  They get strong mode in their tests, and their downstream users, when they opt in, get strong mode in their tests/app as well.

Let me walk through the scenarios for when the user upgrades to the NNBD SDK.

- User has an existing standalone script, some upstream dep is not migrated.
    - User tries to run the script, gets an error saying "you're not ready to migrate, etc" + other static errors
        - User opts the script out, continues as before
        - User passes --no-null-safety, fixes static errors, continues.  They have a constant reminder that they are opting out of null safety in front of them.
- User has an existing standalone script, all upstream deps are migrated
    - User gets static errors, opts out, continues as before
    - User gets static errors, fixes those errors, runs script, gets runtime errors, fixes them
      - If User **really** wants to not fix the runtime errors, there is an override, but they have to pass a flag reminding them that they have opted out.
- User writes a new standalone script
     - Basically the same scenarios as above
- User has an existing package, or writes a new package
     - User does not upgrade the language version, continues as before.
     - User upgrades the language version, some upstream deps not migrated
        - User gets told they can't run unless they opt out, or pass the --no-null-safety flag
        - User opts out, they're fine
        - User migrates anyway and passes the --no-null-safety flag
            - User knows they are opting out of null safety
            - Pub will strongly discourage publishing
    - User upgrades the language version, all upstream deps migrated
        - User fixes static errors
        - Tests are **automatically** run in strong mode, User fixes runtime errors
            - Package is published in the best possible state
        -  User may choose to override and run tests in --no-null-safety mode.
            - Package is published in a state that may not work for downstream users.

It seems to me that in all of these scenarios, defaulting the user into running in strong mode is the "Pit of Success".  By default, they get put on the path that leads them to getting full null safety, and which leads them to making their package ready for their downstream users to get on the path to full null safety.  Any other choice that I see means that the default path for users is to not make their code runtime null safe, which in turn means that their downstream users can't get on the good path, even if they want to.

**Comment by lrhn** [comment]
I agree that it's probably a good choice in the long run, and it's definitely workable.
People who know what they are doing can get what they want.

It's primarily the first-time experience I'm worried about, where a new user writes a program without worrying about "language versioning" and other esoteric concepts. Then they add a dependency, because that's what you do, and then their program stops compiling, telling them that they have to do something about concepts they never considered before.

At *that* point, it is extremely important how we handle and explain the problem.

They might have to *downgrade* their main file to make the program run - that is, rewriting it to not be null safe any more. Anything that forces downgrading is a warning sign to me. Sure, they can move the main file to another library and create a version 2.7 library that does nothing except `import "realmain.dart" as other_main; main() => other_main.main();`, but that's not something we can expect to explain. We can have a quick-fix for it (with appropriate comments in the generated file), but we still risk people keeping the "main.dart" + "realmain.dart" file structure forever.

We can start out spreading the word that you should *not* upgrade your main file. If that's part of the initial documentation that people see, then upgraders will hopefully be aware.

We could (and probably should) make the upgrade tool not upgrade any file with a `main` method unless you pass a flag to force it. (Remembering that all test files have `main` methods).

And we need IDE/analysis-server integration to warn people who write main files that are Null Safe - maybe only if anything in their transitive dependencies are not Null Safe.

That is why I see this as a potentially dangerous approach: The incentives to users are skewed towards silly indirection files or keeping all scripts back - including tests, and I worry about the user experience of hitting the sharp edge.

(My comments on the other thread were about defaults for a single library. Choosing the newest language version absent of any hints is a "safe" choice because it's great if it's right, and if not, you can just write marker for that one file, and then it's solved. The solution can be applied where the problem is detected. Choosing strong mode over weak mode based on the entry library's language version propagates a choice onto the entire program, even though it's not something the entry-point library is actually in control of. You can't have an NNBD main file in a not-completely-migrated program, so the solution is to downgrade the main file or upgrade all dependencies, neither of which are nice.)

**Comment by leafpetersen** [comment]
> It's primarily the first-time experience I'm worried about, where a new user writes a program without worrying about "language versioning" and other esoteric concepts. Then they add a dependency, because that's what you do, and then their program stops compiling, telling them that they have to do something about concepts they never considered before.

Ok, work through this though.  Suppose we take a different approach.  They write a program without worrying about "runtime null safety" and other esoteric concepts.  They write it, run it, ship it.  When do they find out that they're not done, and finish the job.  You're asking **every single user* to have to **explicitly** go through this process of learning that there is a weak mode, and that they have to do something special to get strong mode, and to then fix the code they thought was already fixed.

> They might have to _downgrade_ their main file to make the program run - that is, rewriting it to not be null safe any more.

Or just explicitly run with --no-null-safety, but yes this can happen, it's true.  So we have a small fraction of users having to deal with the weak/strong distinction.  The alternative is that all users have to deal with the weak/strong distinction.  Why is this better?

> We can start out spreading the word that you should _not_ upgrade your main file. If that's part of the initial documentation that people see, then upgraders will hopefully be aware.

I absolutely don't think this is a good idea.

> We could (and probably should) make the upgrade tool not upgrade any file with a `main` method unless you pass a flag to force it. (Remembering that all test files have `main` methods).

Absolutely not.  **The entire point is to get test files opted into strong mode by default.** That's the whole ballgame!  

> That is why I see this as a potentially dangerous approach:

What is your proposed alternative?

> Choosing the newest language version absent of any hints is a "safe" choice because it's great if it's right, and if not, you can just write marker for that one file, and then it's solved. The solution can be applied where the problem is detected.

Right!  Exactly as with my proposal here!

> Choosing strong mode over weak mode based on the entry library's language version propagates a choice onto the entire program, even though it's not something the entry-point library is actually in control of.

I don't understand what you're saying here.  Of course it's under the control of the entry-point library?  That's the whole proposal!

> You can't have an NNBD main file in a not-completely-migrated program, so the solution is to downgrade the main file or upgrade all dependencies, neither of which are nice

That's not true.  This proposal does not take away any options.  You simply can run with `--no-null-safety`.  Easy as pie.  All this does is change what the defaults are.

**Comment by lrhn** [comment]
Maybe I'm over-worrying, but I don't think I am. It's not the technical problems as much as it's the user experience I'm worried about.

My assumption is that if you program is not entirely Null Safe, then you can't run it in strong mode.
If you run the Null-Safety migration tool on your package, it will upgrade every library to Null Safety, even if some of your dependencies are not Null Safe yet.
If you then run `pub run test`, every test will fail to compile. That's a bad user experience.
It means our migration tool just did something that it knew would not work.

You can then manually pass a `-no-null-safety`  argument every time you test until ... some time not under your control. I'd revert my migration to avoid that (we don't have any back-migration tools), even if all my *direct* dependencies are all migrated and my API is known to be correct.

The fix must be proportional to the problem. Having to pass a parameter is not ergonomic for anything except playing around with a new feature. 
Can you make Travis do it? Can you make your IDE do it? Probably, but it's more complication that we are offloading on the user, just for daring to migrate before all transitive dependencies (and dev-dependencies) are migrated.

So, I want an easy way to *override* the default. There is nothing wrong with a default which is at least loosely correlated to the correct answer, but if there are false positives/negatives, there should be a convenient way to choose something other than the default.
I'd seriously consider the "redirecting main file" solution otherwise, and write myself a tool to generate it for me, and that's a sign that we are making users go through hoops to get what they want (well, what I assume they want, because that's what I'd want).

That is: It's OK that the default for choosing strong/weak mode is a based on the language level of the script, as long as there is an *easy* way to override it.
That would mean:
* If the script library has Null Safety enabled (language version >= 2.10 or most recent+experiment flag), then the default is strong mode.
* In that case, a second marker can be used to force weak mode.

The second marker could be a `//@nullsafe=weak` comment at the top of the file or near the `main` method, or a `@pragma("nullsafe:weak")` annotation on the `library` or `main` declaration (we can make a shorthand for that), or something else.

It means that when the migration tool upgrades your library, it checks whether any of your current transitive dependencies are not null safe, and if so, it inserts the weak-mode marker along with a comment saying "Remove this to enable full run-time Null Safety checking.".
(If the upgrade tool cooperates with Pub, it can even check if the depended on packages have any Null Safe version, and encourage you to upgrade first, or Pub could tell you on a Pub get/upgrade that *now* all your dependencies and/or dev-dependencies are Null Safe, so you can remove any weak-mode markers from your app and/or tests).

I argued against a marker in the original thread, but mainly because you can't add a marker to a source that you are not in control of. If we are doing the default *consistently*, then the third-party script is probably already correct about its choice of weak/strong mode, and forcing it to something else is dangerous.
If you do seriously want to anyway, which should be rare, there is the option of a redirecting main, potentially as a `data` URI: 
```dart
Uri redirectingMain(Uri script) => 
    Uri.dataFromString("//@dart=2.7\nimport'$script'as r;main()=>r.main();")
```

So, that brings be back to option 1. of https://github.com/dart-lang/sdk/issues/40632 as a default, but with option 3. as an override only when necessary. That gives both a reasonable default for new code, and an easy way to avoid the default when it's incorrect.

**Comment by kevmoo** [comment]
There are many places here we can help the user, though.

1) The analyzer could report hints libraries with `main` that are opted-in, but have opted-out transitive dependencies
2) The migration tool could also provide such warnings
3) We can make sure the error that is printed on runtime errors is helpful for users – include a short link more details on the website, etc

**Comment by leafpetersen** [comment]
@lrhn Reading your comment above, you seem really focused on making it easy for users to both migrate out of order, and publish the results. I don't think this is a goal - in fact, I think it is an anti-goal.  This is not something I want to make easy.  Possible, yes.  Easy, no.  

> My assumption is that if you program is not entirely Null Safe, then you can't run it in strong mode.

Correct.

> If you run the Null-Safety migration tool on your package, it will upgrade every library to Null Safety, even if some of your dependencies are not Null Safe yet.

We have discussed the tool warning you about this.

> If you then run `pub run test`, every test will fail to compile. That's a bad user experience.

No.  It's a good user experience.  It's telling you that in your package's current state, your downstream users will not be able to run correctly.

> You can then manually pass a `-no-null-safety` argument every time you test until ... some time not under your control. I'd revert my migration to avoid that (we don't have any back-migration tools), even if all my _direct_ dependencies are all migrated and my API is known to be correct.

Good.  You should.  You shouldn't be migrating a pub package before all of the transitive dependencies are migrated.

> Can you make Travis do it? 

I hope not. I don't want people publishing packages that don't have their tests running in strong mode.

> The second marker could be a `//@nullsafe=weak` comment at the top of the file or near the `main` method,

This will, I think, have to be overridable by the flag.  That is, if the user passes --null-safety or --no-null-safety, this has to take precedence, since the modular build systems will need to be driven off of a build configuration and not from parsing the source.  Other than that, I guess this would be ok.  I'm not super keen on there being a way to embed weak execution into the source - it makes it far too easy to just opt out and publish.  And once you've done that **you have blocked everybody downstream of you from correctly migrating**.  As a package, you are better off not migrating than partially migrating.   So the only valid use case I see for this is in apps, and there I don't see why running with the flag is too much to ask.

**Comment by lrhn** [comment]
I'm not sure I care about publishing, but I do care about a user running the migration tool, and then wanting to downgrade some file again. We do not provide a good way to do that. (Sure, it's one git command, but still a bad experience, and there are workarounds.)

It's a *script* file we are talking about. There are no dependencies on *those*, they are intended for direct invocation from the command line or similar. There might be dependencies on the rest of the package, but we are not preventing that from being migrated.

Nothing prevents you from publishing as long as you don't migrate your *scripts*.

I guess not being able to migrate your tests may be a publishing blocker, but that's so indirect that I'm not sure people will see the connection.

As a user, I'd like to be able to migrate my entire package, including tests, and run that in weak mode. If `pub run test` can't do that, our tools are not sufficient (so the `test` package needs to understand the weak-mode flag).

There are workarounds, and as a package author, I'd use them. Not to publish anything, but just to *run*. 

I think we should give the users a *simpler* workaround for the situation where our heuristic default is wrong. Maybe the command line flag is enough, I'm just not convinced that it is. We have to ensure that our entire developer experience works with that. (At least the analyzer doesn't care).

**Comment by jakemac53** [comment]
> I guess not being able to migrate your tests may be a publishing blocker, but that's so indirect that I'm not sure people will see the connection.

You have no idea if your package is actually null safe until you do this. People should really not be publishing without running tests in strong mode.

Passing tests should be an integral part of the publishing process. Once you opt in that should include running them in strong mode. If you can't you shouldn't publish.

**Comment by lrhn** [comment]
Then don't publish. That is not my point. 

My point is that if you do upgrade, even if you don't plan to publish yet, even if you know that some of your dependencies are not migrated yet, but you just want to try migrating anyway - you are stuck. Your test files are migrated, but running them just got much harder.
Migration is not a perfect operation, you may have to do some hand-fiddling to get it right.
Then you *don't* want to have to revert that again.

We have a default that might be wrong. If it is wrong, there should be a *proportional* way to change the default. I don't think a command line flag is adequate, and I'd like a way to change the default in way that is stable and persistent, preferably in the affected file itself. 

Alternatively, we could just refuse to migrate any code which has a non-migrated dependency.
That would at least prevent users from falling into this usability pit.

**Comment by kevmoo** [comment]
> That would at least prevent users from falling into this usability pit.

I think WARNING at a minimum, is a great idea, @lrhn

**Comment by jakemac53** [comment]
@leafpetersen mentioned this above but some configuration in the file itself is not sufficient for all tools (for example bazel). We have to select all our transitive dependencies before reading the file - and in this case part of that selection is choosing either the strong or weak versions of pre-compiled dependencies. That isn't a blocker per say but it does mean it can't be a universally consistent thing.

I think there is also a reasonable argument that hardcoding weak mode for a file could lead to accidental publishing of packages which can only run tests in weak mode. An explicit flag can be annoying but it does keep you aware of what you are doing.

**Comment by leafpetersen** [comment]
> It's a _script_ file we are talking about. There are no dependencies on _those_, they are intended for direct invocation from the command line or similar. There might be dependencies on the rest of the package, but we are not preventing that from being migrated.
> 
> Nothing prevents you from publishing as long as you don't migrate your _scripts_.
> 
> I guess not being able to migrate your tests may be a publishing blocker, but that's so indirect that I'm not sure people will see the connection.

@lrhn I don't understand what you're trying to say with these three paragraphs, could you clarify?  It might help if you quoted what (if anything) you were responding to?


> As a user, I'd like to be able to migrate my entire package, including tests, and run that in weak mode. If `pub run test` can't do that, our tools are not sufficient (so the `test` package needs to understand the weak-mode flag).

I believe that is the plan, yes.

**Comment by lrhn** [comment]
What I'm asking for is an increment on the proposed feature, to be used in (so far entirely hypothetical) cases where the default is not what the user wants.

So, rather than block progress (especially since I'm vacating for Easter early), let's implement the "entry point language version decides strong mode/weak mode" plus command-line override flag first.
If it turns out that it is the usability problem that I worry about, we can add the in-file override flag at a later time (or some other similar feature).

I do recommend that the migration tool consider whether it should be migrating scripts, or that we ensure that the user-experience is generally pleasant when scripts are migrated (can I run tests from inside VSCode?)

**Comment by leafpetersen** [comment]
> What I'm asking for is an increment on the proposed feature, to be used in (so far entirely hypothetical) cases where the default is not what the user wants.
> 
> So, rather than block progress (especially since I'm vacating for Easter early), let's implement the "entry point language version decides strong mode/weak mode" plus command-line override flag first.
> If it turns out that it is the usability problem that I worry about, we can add the in-file override flag at a later time (or some other similar feature).
> 
> I do recommend that the migration tool consider whether it should be migrating scripts, or that we ensure that the user-experience is generally pleasant when scripts are migrated (can I run tests from inside VSCode?)

This all seems reasonable to me.

**Comment by leafpetersen** [comment]
@mit-mit This is the discussion that I referenced earlier today in which @lrhn was advocating for a declarative way to opt an entry point into weak execution.  See specifically the comment above this one.

For context, @mit-mit and I discussed today the possibility of doing this via a marker in the pubspec which would get mapped to a package_config.json entry which would essentially change the default behavior for all entry points in the package.  The advantage of this is that it is something that pub can easily recognize and surface to users.  

For the time being, we proposed to take a wait and see approach to see if this becomes an issue.
